Create A Switch Block Immediately After Upgrade or Genesis

xcthulhu opened the pull request:
After either genesis or an upgrade, we immediately create a switch block.

Closes #1666

Commit:
Create switch block immediately after genesis or upgrade

# Conflicts:
#	node/src/components/contract_runtime.rs
#	node/src/components/storage.rs
#	node/src/reactor/participating.rs

Commit:
Not getting blocks before emergency restart from storage

Commit:
Getting rid of commented out code

Commit:
Don't get blocks before emergency restart

Commit:
Proper minimum era

Commit:
No longer exposing genesis validators

goral09 commented on types/src/system/auction/mod.rs:
```diff
@@ -506,10 +506,6 @@ pub trait Auction:
         let base_round_reward = self.read_base_round_reward()?;
         let era_id = detail::get_era_id(self)?;
 
-        if reward_factors.keys().ne(seigniorage_recipients.keys()) {
-            return Err(Error::MismatchedEraValidators);
-        }
```
Why are you removing this check? This was added after we had a situation/bug where those did indeed not match and node later panic'd.

xcthulhu replied:
When we construct a block after genesis/upgrade we make the reward factors map empty.

goral09 replied:
Then maybe we should be checking if it's a subset of the expected validator set? Rather than removing the check althogether. It's not strictly necessary but I think by removing it we're losing more that we could gain if we fit it into the new situation.

afck replied:
If it's not a subset that will already cause a `ValidatorNotFound` error in line 515.

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -256,6 +265,7 @@ where
         (self
             .protocol_config
             .last_activation_point
+            .saturating_add(1)
```
Please add a comment (here and below, in `iter_past_other`) why we're adding 1 era to this. No one will remember that in a week.

afck replied:
One week later, trying to remember… :sweat_smile: :thinking: 

The activation point era itself contains only the single switch block we created after the upgrade. There is no consensus instance for it.

xcthulhu replied:
Added a clarifying comment in f377bdc7a

goral09 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -457,58 +467,30 @@ where
+            let start_height = key_block.height() + 1;
+            let era_start_time = key_block.timestamp();
+            let seed = Self::era_seed(*booking_block_hash, key_block.accumulated_seed());
+            // If it's neither genesis nor upgrade nor restart, we use the validators from
+            // the key block and ban validators that were faulty in previous eras.
```
This comment doesn't make as much sense anymore as it used to, as from the perspective of `EraSupervisor` it's never a _genesis or upgrade or restart_ since we're not creating an additional switch block when it _is_ one of these situations. We could update the comment to reflect that otherwise it's easily missed - it's created in a constructor of participating reactor which is far from here and developers of consensus component shouldn't even have to know about it.

xcthulhu replied:
Removed in 69fcec87f

goral09 commented on node/src/components/linear_chain_sync/operations.rs:
```diff
@@ -312,6 +276,72 @@ where
+    }
+}
+
+/// Get trusted switch block; returns `None` if we are still in the first era.
```
```suggestion
/// Get trusted switch block.
/// Returns `None` if we are still in the first era.
```

Commit:
LinearChainSync logic assumes switch block at genesis/upgrade

# Conflicts:
#	node/src/components/linear_chain_sync/error.rs
#	node/src/components/linear_chain_sync/operations.rs

fizyk20 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -180,45 +183,51 @@ where
+                .last_emergency_restart
+                .unwrap_or_else(|| EraId::from(0u64));
+            let key_block_era_id = era_id.saturating_sub(1);
+            if key_block_era_id < minimum_era {
+                continue;
+            }
```
This shouldn't even be possible, should it? `era_ids` shouldn't contain an era ID that's less than `minimum_era + 1` because `iter_past()` only starts at `last_emergency_restart + 1` (or 1 if there was no emergency restart).

xcthulhu replied:
The old async code had the check so I just replicated that...

afck replied:
Also, `handle_initialize_eras` will panic if a key block is missing. Better return an error here instead.
(Or panic? Since it can be verified within this module that it's unreachable?)

xcthulhu replied:
Removed this check in e3a2e078a

fizyk20 commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -180,45 +183,51 @@ where
+                .read_switch_block_header_by_era_id(key_block_era_id)?
+                .ok_or_else(|| anyhow::Error::msg("No such switch block"))?;
+            key_blocks.insert(era_id, block_header);
+            let booking_block_era_id = era_id
+                .saturating_sub(era_supervisor.protocol_config.auction_delay)
+                .saturating_sub(1);
+            if booking_block_era_id < minimum_era {
+                continue;
             }
```
Have you tried running an emergency restart test? Because I'm pretty sure this would cause it to fail.

Consider this scenario: we are starting right after an emergency restart, the last era before was `N-1`, now we created a dummy switch block for era `N` and are trying to initialize era `N+1`. `minimum_era` will be `N`. `booking_block_era_id` will be `N - auction_delay`, so it's less than `minimum_era`, so we'll hit the `continue` statement and won't save a booking block for era `N+1`. But then in `handle_initialize_eras` we panic if there's no booking block for an era - which means we'll panic.

Am I missing something?

afck replied:
Good catch! We need to use `BlockHash::default()` instead of `continue`.

xcthulhu replied:
Fixed in c2fa0b4b5

fizyk20 reviewed (commented):
Just a quick partial review - I'll finish it sometime later today.

Commit:
Small network no longer queries storage or the contract runtime for next era valdiators

dwerner commented on node/src/reactor/participating.rs:
```diff
@@ -539,20 +540,150 @@ impl reactor::Reactor for Reactor {
+                    // Unless it's an emergency upgrade, if there's a block higher than ours we will
+                    // over-write an existing blockchain if we don't exit immediately.
+                    if chainspec.protocol_config.last_emergency_restart
+                        == Some(upgrade_block_header.era_id())
```
Is this correct? This seems to ask if we are currently in an emergency upgrade, not `unless it's an emergency upgrade`. Perhaps the comment should be "If we are **not** doing an emergency upgrade, and if there's a block higher than ours..."

dwerner replied:
```suggestion
                        != Some(upgrade_block_header.era_id())
```

Commit:
Inlining initial era

goral09 commented on node/src/components/small_network.rs:
```diff
@@ -1028,22 +960,18 @@ where
-                    Effects::new()
-                }
+                let _era_id = block.header().era_id();
+                if let Some(next_era_validators) =
+                    block.take_header().maybe_take_next_era_validator_weights()
+                {
+                    let upcoming_validators: HashSet<PublicKey> =
+                        next_era_validators.into_keys().collect();
+                    self.outgoing_limiter
+                        .update_validators(Default::default(), upcoming_validators.clone());
+                    self.incoming_limiter
+                        .update_validators(Default::default(), upcoming_validators);
```
This is now changing the logic. Previously, on every switch block it would detect upcoming validators for 2 eras in the future (the immediate era after the switch block and validators after that era, since Auction contract has already determined who those validators will be). New logic only updates the next era validators so we don't get the benefit of prioritizing soon-to-be-validators nodes' traffic. The feature was/is even more important if we increase the `auction_delay` to more than 1 - that means we can prioritize nodes that will be validators in 2,3,... eras.  /cc @marc-casperlabs

marc-casperlabs replied:
@goral09 is correct. However that was the *intended* behavior. Sadly, in the past, the function that is supposed to determine the `n+2` era validators would always return an empty set due to a bug/shortcoming in the function itself, but fixing that bug was outside the scope of implementing the upstream limiter itself.

So the code in small networking was written in the hopes of that bug being fixed and "future proof" in a way, but as far as I remember the fix has not materialized yet.

We are about to change this section of code once again, so I am fine with the changes here for now. We will replace it before 1.4, and additionally it should not degrade the _actual_ behavior at the moment.

Pretty good catch though nonetheless :+1:

goral09 replied:
Ok. Thanks for responding.

fizyk20 replied:
@marc-casperlabs #1469 was supposed to be the fix you mention :wink:

xcthulhu replied:
This will be handled in an upcoming PR

goral09 reviewed (dismissed):
The change to small network, albeit simplifying the code, is changing the logic of the feature we added to prioritize traffic between validator nodes.

Commit:
Setting initial validators

xcthulhu committed:
Getting rid of semi-colon

Co-authored-by: Mateusz Górski <goral09@users.noreply.github.com>

Commit:
Removing passive voice

xcthulhu commented on node/src/reactor/participating.rs:
```diff
@@ -539,55 +540,181 @@ impl reactor::Reactor for Reactor {
+                            return Err(Error::NonEmergencyUpgradeWillClobberExistingBlockChain {
+                                preexisting_block_header: Box::new(preexisting_block_header),
+                            });
+                        }
```
Chatting with @Fraser999 , we should always do this

WDYT @fizyk20 ?

goral09 replied:
What if someone misses to schedule an upgrade on time, and its node already managed to build blocks in an era where it should be running a new version? Node operator would try to fix its mistake by scheduling the upgrade again and IIRC, it would try to roll back those blocks. Is it possible to do without deleting them from the storage? In which case, such node would fail with the above error.

xcthulhu replied:
In the circumstance you describe the code would never get here.  It would set `latest_block_header` in the `if let` branch at line 555 (see [here][1]).

[1]: https://github.com/casper-network/casper-node/pull/1796/files#diff-212250b2151062263c01ec4511602e2ada0b8a83dd02b8727589737f3d3312ebR555

fizyk20 replied:
What should we always do? Bail if there exists a block higher than the current activation point? I'm not sure. I imagine we might _want_ to overwrite existing blocks sometimes, like when we're doing an emergency upgrade... though we'd probably use a hard reset in such cases and the offending blocks should be removed earlier, right?
Also, when would we _not_ have a `latest_block_header` if there are blocks in storage already?

xcthulhu replied:
> Also, when would we not have a `latest_block_header` if there are blocks in storage already?

The joiner reactor might not have the latest block header if it was skipped (ie, in the case of an upgrade or genesis).

xcthulhu commented on node/src/reactor/participating.rs:
```diff
@@ -539,55 +540,181 @@ impl reactor::Reactor for Reactor {
+                        Some(chainspec.core_config.auction_delay),
+                        Some(chainspec.core_config.locked_funds_period.millis()),
+                        Some(chainspec.core_config.round_seigniorage_rate),
+                        Some(chainspec.core_config.unbonding_delay),
```
Chatting with @Fraser999 , consider having `CoreConfig` as an argument.

goral09 replied:
There's a lot of ngineering practices advising against doing that:
* you should be passing only what's needed - that way it makes it easier to test (one doesn't have to mock all other fields); you make it easier to extend (you're not hardcoding the structure to use the config struct), it's more flexible (if we refactor `CoreConfig` we would have to also refactor `CommitUpgrade`.
* you won't be copying as much data (rather than a couple of fields, it would need to copy whole structure).
* `CoreConfig` is a type from "another level" of the stack, different context you might say - it's a configuration layer rather than EE specific thing. Which is related to the first point - they might be evolving differently, one can be an "interpretation" of the other (derived from), etc.

xcthulhu replied:
The code was adapted from some of Fraser's ChainSpec code [here][1].  We identified it as having too many arguments and proposed it take a data structure instead.

[1]: https://github.com/xcthulhu/casper-node/blob/1c1d2564d40d0e72472ca16cab9bb35fe180245d/node/src/components/chainspec_loader.rs#L534-L547

afck commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -457,58 +467,30 @@ where
     ) -> HashMap<EraId, ProtocolOutcomes<I, ClContext>> {
         let mut result_map = HashMap::new();
 
         for era_id in self.iter_past(self.current_era, self.bonded_eras().saturating_mul(2)) {
```
We could probably just iterate over the `key_blocks` directly now.

xcthulhu replied:
Fixed in 4206c1261

afck commented on node/src/components/consensus/era_supervisor.rs:
```diff
@@ -854,11 +814,11 @@ where
         key_blocks: HashMap<EraId, BlockHeader>,
         booking_blocks: HashMap<EraId, BlockHash>,
-        validators: BTreeMap<PublicKey, U512>,
+        _validators: BTreeMap<PublicKey, U512>,
```
That can be removed now.

xcthulhu replied:
Fixed in 4206c1261

afck commented on node/src/components/linear_chain_sync/operations.rs:
```diff
@@ -217,7 +180,7 @@ pub(crate) fn check_sufficient_finality_signatures(
         .sum();
 
-    let lower_bound = Ratio::from(1u64) - finality_threshold_fraction;
+    let lower_bound = (Ratio::from(1u64) + finality_threshold_fraction) / Ratio::from(2u64);
```
```suggestion
    let lower_bound = (finality_threshold_fraction + 1) / 2;
```

xcthulhu replied:
Fixed in e36dc7b85

afck commented on node/src/components/linear_chain_sync/operations.rs:
```diff
@@ -312,6 +276,73 @@ where
     Ok(None)
 }
 
+pub(super) struct SwitchBlockInfo {
```
I'd call this `KeyBlockInfo`, so it's clearer that the fields refer to the _next_ era.
(Alternatively, rename the fields to `next_era_validator_weights` and `timestamp`.)

xcthulhu replied:
Fixed in a7a3a3118

afck commented on node/src/components/linear_chain_sync/operations.rs:
```diff
@@ -312,6 +276,73 @@ where
+pub(super) struct SwitchBlockInfo {
+    validator_weights: BTreeMap<PublicKey, U512>,
+    era_start: Timestamp,
+    switch_block_height: u64,
```
Maybe just `height`? (Seems redundant.)
```suggestion
    height: u64,
```

xcthulhu replied:
Fixed in a7a3a3118

afck commented on node/src/components/linear_chain_sync/operations.rs:
```diff
@@ -312,6 +276,73 @@ where
+}
+
+/// Get trusted switch block.
+/// Returns `None` if we are still in the first era.
```
```suggestion
/// Returns an error if we are still in the first era, or the era after an emergency restart.
```

afck commented on node/src/components/small_network.rs:
```diff
@@ -1028,22 +970,17 @@ where
+                    let upcoming_validators: HashSet<PublicKey> =
+                        next_era_validators.into_keys().collect();
+                    self.outgoing_limiter
+                        .update_validators(Default::default(), upcoming_validators.clone());
```
Shouldn't we treat these as _active_ validators now instead of upcoming? After all, that era is starting right now.

```suggestion
                        .update_validators(upcoming_validators.clone(), Default::default());
```

xcthulhu replied:
As discussed, @marc-casperlabs plans to take care of this in an upcoming PR.

afck replied:
Still, wouldn't it be better to use the first argument here instead of the second? (Not sure what difference it makes…?)

afck commented on node/src/components/small_network.rs:
```diff
@@ -327,6 +322,15 @@ where
+            component
+                .incoming_limiter
+                .update_validators(Default::default(), initial_validators);
+        }
```
These should probably be treated as _active_, not upcoming, i.e. the first argument of the limiter. (See comment below.)

xcthulhu replied:
I believe @marc-casperlabs intended to take care of this in an upcoming PR.  After this PR, I have a follow-up to clean up this code a bit on the ContractRuntime end.

afck commented on node/src/reactor/participating.rs:
```diff
@@ -539,55 +540,181 @@ impl reactor::Reactor for Reactor {
+            let finalized_block;
+
+            let chainspec = chainspec_loader.chainspec();
+            match chainspec_loader
+                .chainspec()
```
```suggestion
            match chainspec
```

xcthulhu replied:
Fixed in e5a8fdbca

afck commented on node/src/reactor/participating.rs:
```diff
@@ -539,55 +540,181 @@ impl reactor::Reactor for Reactor {
+                ActivationPoint::Genesis(genesis_timestamp) => {
+                    // Check that no blocks exist in storage so we don't try to run genesis again on
+                    // an existing blockchain
+                    if let Some(first_block_header) = storage.read_block_header_by_height(1)? {
```
```suggestion
                    if let Some(first_block_header) = storage.read_block_header_by_height(0)? {
```

afck commented on node/src/reactor/participating.rs:
```diff
@@ -539,55 +540,181 @@ impl reactor::Reactor for Reactor {
+                            });
+                        }
+                    }
+                    let chainspec = chainspec_loader.chainspec();
```
I think `chainspec` is still in scope.

```suggestion
```

afck commented on node/src/reactor/participating.rs:
```diff
@@ -539,55 +540,181 @@ impl reactor::Reactor for Reactor {
+                        Some(chainspec.core_config.unbonding_delay),
+                        global_state_update,
+                    ))?;
+                    info!("upgrade chainspec name {}", chainspec.network_config.name);
+                    info!("upgrade state root hash {}", post_state_hash);
```
```suggestion
                    info!(
                        network_name = %chainspec.network_config.name,
                        %post_state_hash,
                        "upgrade committed"
                    );
```

afck commented on node/src/reactor/participating/error.rs:
```diff
@@ -39,4 +44,50 @@ pub enum Error {
     Serialization(#[source] bincode::ErrorKind),
+
+    /// Engine state error.
+    #[error("engine state: {0}")]
```
We could use `transparent` for this and the next two.

```suggestion
    #[error(transparent)]
```

afck commented on node/src/reactor/participating/error.rs:
```diff
@@ -39,4 +44,50 @@ pub enum Error {
+    /// Cannot run genesis on pre-existing blockchain.
+    #[error("Cannot run genesis on pre-existing blockchain. First block: {first_block_header:?}")]
+    CannotRunGenesisOnPreExistingBlockchain {
+        /// The first block header.  Should have height 1.
```
```suggestion
        /// The first block header.  Should have height 0.
```

afck commented on node/src/reactor/participating/error.rs:
```diff
@@ -39,4 +44,50 @@ pub enum Error {
+        upgrade_era_id: EraId,
+    },
+
+    /// Non-emergency upgrade will clobber existing block-chain.
+    #[error(
+        "Non-emergency upgrade will clobber existing block-chain. \
```
```suggestion
    /// Non-emergency upgrade will clobber existing blockchain.
    #[error(
        "Non-emergency upgrade will clobber existing blockchain. \
```

afck commented on node/src/reactor/participating/error.rs:
```diff
@@ -39,4 +44,50 @@ pub enum Error {
+    #[error("bytesrepr error: {0}")]
+    BytesReprError(#[from] bytesrepr::Error),
+
+    /// Cannot run genesis on pre-existing blockchain.
+    #[error("Cannot run genesis on pre-existing blockchain. First block: {first_block_header:?}")]
```
```suggestion
    /// Cannot run genesis on preexisting blockchain.
    #[error("Cannot run genesis on preexisting blockchain. First block: {first_block_header:?}")]
```

(Or use "pre-existing" below, too.)

Commit:
Adding comments about missing consensus for activation point era

Commit:
Removing spurious comment

Commit:
Key blocks are always less than minimum era, skipping check

Commit:
Use BlockHash::default() instead of continue

Commit:
Changing era iteration to traverse a BTreeMap

Commit:
Moved check_sufficient_finality_signatures to utils

 - Simplified arithmetic

Commit:
Removing dead code

Commit:
SwitchBlockInfo -> KeyBlockInfo

xcthulhu committed:
Fixing comment

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

Commit:
Simplifying match

xcthulhu committed:
block-chain -> blockchain

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

xcthulhu committed:
Simplifying error message

Co-authored-by: Andreas Fackler <afck@users.noreply.github.com>

Commit:
Already have `chainspec` in scope

fizyk20 reviewed (approved):
Good to merge, outstanding issues have to do with re-planning emergency upgrades for fast sync, which we agreed to do separately.

dwerner reviewed (approved):
Had a read and looks good to me.
